### PR TITLE
prov/efa: atomic and medium message protocol

### PR DESCRIPTION
**Pull request by wzamazon**
This PR contains 3 commits:

1. implement atomic
2. remove RTS packet type
3. implement medium message protocol

### CODE REVIEW

**Comment by wzamazon** [comment]
Did check patch in latest revision.

File: prov/efa/src/rxr/rxr_pkt_type_req.c [hunk 1]
```diff
@@ -134,6 +139,10 @@ size_t rxr_pkt_req_base_hdr_size(struct rxr_pkt_entry *pkt_entry)
 	else if (base_hdr->type == RXR_SHORT_RTR_PKT ||
 		 base_hdr->type == RXR_LONG_RTR_PKT)
 		hdr_size += rxr_get_rtr_hdr(pkt_entry->pkt)->rma_iov_count * sizeof(struct fi_rma_iov);
+	else if (base_hdr->type == RXR_WRITE_RTA_PKT ||
+		 base_hdr->type == RXR_FETCH_RTA_PKT ||
+		 base_hdr->type == RXR_COMPARE_RTA_PKT)
+		hdr_size += rxr_get_rta_hdr(pkt_entry->pkt)->rma_iov_count*sizeof(struct fi_rma_iov);
```
Comment (rwespetal) [inline prov/efa/src/rxr/rxr_pkt_type_req.c]: Spacing is off here

File: prov/efa/src/rxr/rxr_atomic.c [hunk 2]
```diff
@@ -0,0 +1,477 @@
...
+		memcpy(&shm_msg, msg, sizeof(struct fi_msg_atomic));
+		shm_msg.addr = peer->shm_fiaddr;
+		return fi_compare_atomicmsg(rxr_ep->shm_ep, &shm_msg,
+					    comparev, compare_desc, compare_count,
+					    resultv, result_desc, result_count,
+					    flags);
+	}
+
+	ofi_ioc_to_iov(resultv, atomic_ex.resp_iov, result_count, datatype_size);
+	atomic_ex.resp_iov_count = result_count;
```
[the remark this answers is not in the document]
Comment (wzamazon) [reply prov/efa/src/rxr/rxr_atomic.c]: See my previous comments.

File: prov/efa/src/rxr/rxr_atomic.c [hunk 3]
```diff
@@ -0,0 +1,477 @@
...
+	if (rxr_env.enable_shm_transfer && peer->is_local) {
+		memcpy(&shm_msg, msg, sizeof(struct fi_msg_atomic));
+		shm_msg.addr = peer->shm_fiaddr;
+		return fi_fetch_atomicmsg(rxr_ep->shm_ep, &shm_msg,
+					  resultv, result_desc, result_count,
+					  flags);
+	}
+
+	ofi_ioc_to_iov(resultv, atomic_ex.resp_iov, result_count, datatype_size);
+	atomic_ex.resp_iov_count = result_count;
```
[the remark this answers is not in the document]
Comment (rwespetal) [reply prov/efa/src/rxr/rxr_atomic.c]: I'm not sure I understand, this seems to be using this variable in the context of an iov.

```
void rxr_pkt_handle_atomrsp_recv(struct rxr_ep *ep,
                                 struct rxr_pkt_entry *pkt_entry)
{
...
        ofi_copy_to_iov(tx_entry->atomic_ex.resp_iov,
                        tx_entry->atomic_ex.resp_iov_count,
                        0, atomrsp_pkt->data,
                        atomrsp_hdr->seg_size);
```

File: prov/efa/src/rxr/rxr_pkt_type_req.c (hunk 1 above)
Comment (rwespetal) [reply prov/efa/src/rxr/rxr_pkt_type_req.c]: Sorry - meant spacing around `*`.

File: prov/efa/src/rxr/rxr.h [hunk 4]
```diff
@@ -783,6 +811,23 @@ static inline int rxr_need_sas_ordering(struct rxr_ep *ep)
 		rxr_env.enable_sas_ordering);
 }
 
+static inline int rxr_need_atomic_ordering(struct rxr_ep *ep)
+{
+	/*
+	 * RxR needs to reorder RTS packets for send-after-send guarantees
+	 * only if the application requested it and the core provider does not
```
[the remark this answers is not in the document]
Comment (wzamazon) [reply prov/efa/src/rxr/rxr.h]: Ah, I missed that. Thank you! fixed.

File: prov/efa/src/rxr/rxr_atomic.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (wzamazon) [reply prov/efa/src/rxr/rxr_atomic.c]: ```result``` is a vector of ioc, ```result_count``` is how many ioc in ```result```.

```ofi_ioc_to_iov``` convert ```result``` from a vector of ioc to a vector of iov named ```resp_iov```. ```resp_iov_count``` is number of iov in ```resp_iov```, so it should be set to ```result_count```.

[the remark this answers is not in the document]
Comment (rwespetal) [reply prov/efa/src/rxr/rxr_atomic.c]: Ah, I see it now... sorry for the confusion on my part..

File: prov/efa/src/rxr/rxr_msg.c [hunk 5]
```diff
@@ -58,7 +58,69 @@
...
+	assert(RXR_MEDIUM_MSGRTM_PKT + 1 == RXR_MEDIUM_TAGRTM_PKT);
+
+	int tagged;
+	size_t max_rtm_data_size;
+	ssize_t err;
+	struct rxr_peer *peer;
+
+	assert(tx_entry->op == ofi_op_msg || tx_entry->op == ofi_op_tagged);
+	tagged = (tx_entry->op == ofi_op_tagged);
+	assert(tagged == 0 || tagged == 1);
```
Comment (rwespetal) [inline prov/efa/src/rxr/rxr_msg.c]: This assert can be dropped.

**Comment by wzamazon** [comment]
bot:aws:retest

File: prov/efa/src/rxr/rxr_atomic.c [hunk 6]
```diff
@@ -0,0 +1,477 @@
+#include <ofi_atomic.h>
```
Comment (rajachan) [inline prov/efa/src/rxr/rxr_atomic.c]: This file is missing a copyright header.

File: prov/efa/src/rxr/rxr_atomic.h [hunk 7]
```diff
@@ -0,0 +1,46 @@
+/*
+ * Copyright (c) 2019 Amazon.com, Inc. or its affiliates.
```
Comment (rajachan) [inline prov/efa/src/rxr/rxr_atomic.h]: Year->2020.

File: prov/efa/src/rxr/rxr_pkt_type.h [hunk 8]
```diff
@@ -65,10 +65,11 @@
 #define RXR_READRSP_PKT		5
 #define RXR_RMA_CONTEXT_PKT	6
 #define RXR_EOR_PKT		7
+#define RXR_ATOMRSP_PKT         8
```
Comment (rajachan) [inline prov/efa/src/rxr/rxr_pkt_type.h]: Is this not part of v4 too, and aren't these PKT numbers append-only.

File: prov/efa/src/rxr/rxr_pkt_type_misc.c [hunk 9]
```diff
@@ -515,3 +515,78 @@ void rxr_pkt_handle_eor_recv(struct rxr_ep *ep,
...
+
+	/* rx_entry->atomrsp_buf was filled in rxr_pkt_handle_req_recv() */
+	memcpy((char*)pkt_entry->pkt + RXR_ATOMRSP_HDR_SIZE, rx_entry->atomrsp_buf, atomrsp_hdr->seg_size);
+	pkt_entry->pkt_size = RXR_ATOMRSP_HDR_SIZE + atomrsp_hdr->seg_size;
+	return 0;
+}
+
+void rxr_pkt_handle_atomrsp_sent(struct rxr_ep *ep, struct rxr_pkt_entry *pkt_entry)
+{
+}
```
Comment (rajachan) [inline prov/efa/src/rxr/rxr_pkt_type_misc.c]: What's happening here? I saw a comment elsewhere about releasing a packet entry here.

File: prov/efa/src/rxr/rxr_pkt_type_req.c [hunk 10]
```diff
@@ -1230,3 +1251,244 @@ void rxr_pkt_handle_rtr_recv(struct rxr_ep *ep, struct rxr_pkt_entry *pkt_entry)
...
+void rxr_pkt_handle_write_rta_send_completion(struct rxr_ep *ep, struct rxr_pkt_entry *pkt_entry)
+{
+	struct rxr_tx_entry *tx_entry;
+
+	tx_entry = (struct rxr_tx_entry *)pkt_entry->x_entry;
+	rxr_cq_handle_tx_completion(ep, tx_entry);
+}
+
+void rxr_pkt_handle_fetch_rta_send_completion(struct rxr_ep *ep, struct rxr_pkt_entry *pkt_entry)
+{
```
Comment (rajachan) [inline prov/efa/src/rxr/rxr_pkt_type_req.c]: If we have no action to take, why call an empty function?

File: prov/efa/src/rxr/rxr_atomic.c (hunk 6 above)
Comment (wzamazon) [reply prov/efa/src/rxr/rxr_atomic.c]: added. Thank you!

File: prov/efa/src/rxr/rxr_atomic.h (hunk 7 above)
Comment (wzamazon) [reply prov/efa/src/rxr/rxr_atomic.h]: Fixed, thank you!

File: prov/efa/src/rxr/rxr_pkt_type.h (hunk 8 above)
Comment (wzamazon) [reply prov/efa/src/rxr/rxr_pkt_type.h]: Removed the confusing comments

File: prov/efa/src/rxr/rxr_pkt_type_misc.c (hunk 9 above)
Comment (wzamazon) [reply prov/efa/src/rxr/rxr_pkt_type_misc.c]: I fixed that comment, they are released in rxr_pkt_handle_atomrsp_send_completion()

File: prov/efa/src/rxr/rxr_pkt_type_req.c (hunk 10 above)
Comment (wzamazon) [reply prov/efa/src/rxr/rxr_pkt_type_req.c]: Removed the function and added comments in rxr_pkt_cmd.c.

**Comment by rajachan** [comment]
bot:aws:retest
(Ubuntu repo's seemed temporarily available. Let's see if it was a transient issue)

**Comment by wzamazon** [comment]
Finally, all tests passed! @rajachan

**Pull request merged**